File: packages/maven/src/main/java/io/envy/Envy.java
package io.envy;

import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Loads and validates {@code envy.yaml} configuration natively inside any JVM
 * application — no CLI required. Mirrors the precedence of the envy binary:
 *
 * <pre>
 * System environment → envy.local.&lt;branch&gt;.yaml → envy.local.yaml → schema default → generated mock
 * </pre>
 *
 * <p>Usage:</p>
 * <pre>{@code
 * Map<String, String> config = Envy.load();
 * String dbUrl = config.get("DATABASE_URL");
 * }</pre>
 *
 * <p>Spring Boot: register it as a bean or call it from a
 * {@code @ConfigurationProperties} initializer.</p>
 */
public final class Envy {

    /** Thrown with every problem collected in a single pass. */
    public static final class EnvyException extends RuntimeException {
        public final List<String> problems;

        EnvyException(List<String> problems) {
            super("envy configuration invalid (" + problems.size() + " problem(s)):\n  - "
                    + String.join("\n  - ", problems));
            this.problems = problems;
        }
    }

    private static final String SCHEMA_FILE = "envy.yaml";
    private static final List<String> KNOWN_SCHEMES = List.of(
            "postgresql", "postgres", "mysql", "mariadb", "mssql", "mongodb",
            "mongodb+srv", "redis", "rediss", "amqp", "rabbitmq", "kafka",
            "http", "https", "ws", "wss", "grpc", "ftp", "sftp", "ssh",
            "smtp", "s3", "gs", "azblob", "sqlite");

    private static volatile Map.Entry<Path, Map<String, String>> cached;

    private Envy() {
    }

    /** Load from the current working directory (searching upward). */
    public static Map<String, String> load() {
        return load(Paths.get("").toAbsolutePath());
    }

    /** Load searching upward from {@code startDir}. Results are cached per process. */
    @SuppressWarnings("unchecked")
    public static Map<String, String> load(Path startDir) {
        Path schemaPath = findUpward(SCHEMA_FILE, startDir.toAbsolutePath());
        if (schemaPath == null) {
            throw new EnvyException(List.of(
                    "no " + SCHEMA_FILE + " found in " + startDir + " or any parent directory"));
        }

        Map<String, String> memo = cached == null ? null : cached.getValue();
        if (memo != null && cached.getKey().equals(schemaPath)) {
            return memo;
        }

        Path baseDir = schemaPath.getParent();

        Yaml yaml = new Yaml();
        Map<String, Object> schemaDoc = readYaml(yaml, schemaPath);
        if (schemaDoc == null) {
            throw new EnvyException(List.of(SCHEMA_FILE + ": empty or invalid schema"));
        }
        Object configObj = schemaDoc.get("config");
        Map<String, Object> config = configObj instanceof Map ? (Map<String, Object>) configObj : Map.of();

        String branch = currentBranch(baseDir);
        Path overlayFile = branch == null ? null
                : baseDir.resolve("envy.local." + sanitizeBranch(branch) + ".yaml");
        Path localFile = baseDir.resolve("envy.local.yaml");

        Map<String, Object> overlay = overlayFile != null && Files.exists(overlayFile)
                ? valuesOf(readYaml(yaml, overlayFile), overlayFile.getFileName().toString())
                : new LinkedHashMap<>();
        Map<String, Object> local = Files.exists(localFile)
                ? valuesOf(readYaml(yaml, localFile), "envy.local.yaml")
                : new LinkedHashMap<>();

        List<String> problems = new ArrayList<>();
        Map<String, String> resolved = new LinkedHashMap<>();
        Set<String> sources = new LinkedHashSet<>();

        for (Map.Entry<String, Object> entry : config.entrySet()) {
            String key = entry.getKey();
            Map<String, Object> spec = entry.getValue() instanceof Map
                    ? (Map<String, Object>) entry.getValue()
                    : new LinkedHashMap<>();

            String placed = null;
            String source = null;

            String envValue = System.getenv(key);
            if (envValue != null) {
                placed = envValue;
                source = "env";
            } else if (overlay.containsKey(key)) {
                placed = scalarToString(overlay.get(key), key, problems);
                source = "overlay";
            } else if (local.containsKey(key)) {
                placed = scalarToString(local.get(key), key, problems);
                source = "local";
            } else if (spec.containsKey("default") && spec.get("default") != null) {
                placed = scalarToString(spec.get("default"), key, problems);
                source = "default";
            } else if (Boolean.TRUE.equals(asBool(spec.get("mock")))) {
                placed = mockValue(key);
                source = "mock";
            }

            if (placed == null) {
                if (Boolean.TRUE.equals(asBool(spec.get("required")))) {
                    problems.add("missing required variable " + key);
                }
                continue;
            }

            checkType(spec, key, placed, problems);
            checkFormat(spec, key, placed, problems);
            resolved.put(key, placed);
            sources.add(source);
        }

        Set<String> known = config.keySet();
        for (Map.Entry<String, Map<String, Object>> layer : List.of(
                Map.entry("branch overlay", overlay),
                Map.entry("envy.local.yaml", local))) {
            for (String key : layer.getValue().keySet()) {
                if (!known.contains(key)) {
                    String suggestion = bestKeyMatch(key, known);
                    problems.add(key + " is set in " + layer.getKey()
                            + " but not declared in " + SCHEMA_FILE + " (typo?)"
                            + (suggestion != null ? " — did you mean " + suggestion + "?" : ""));
                }
            }
        }

        if (!problems.isEmpty()) {
            throw new EnvyException(problems);
        }

        Map<String, String> result = Map.copyOf(resolved);
        cached = Map.entry(schemaPath, result);
        return result;
    }

    /** Convenience accessor over {@link #load()}. */
    public static String get(String key) {
        return load().get(key);
    }

    // ---------- internals ----------

    @SuppressWarnings("unchecked")
    private static Map<String, Object> valuesOf(Map<String, Object> doc, String label) {
        if (doc == null) {
            return new LinkedHashMap<>();
        }
        Object inner = doc.get("values");
        if (inner instanceof Map) {
            return (Map<String, Object>) inner;
        }
        if (!doc.isEmpty()) {
            boolean looksFlat = doc.keySet().stream().allMatch(k -> k.equals(k.toUpperCase()) && !k.contains(":"));
            if (looksFlat && !doc.containsKey("service") && !doc.containsKey("version")) {
                return doc;
            }
        }
        throw new EnvyException(List.of(label + ": expected a `values:` mapping"));
    }

    private static Map<String, Object> readYaml(Yaml yaml, Path file) {
        try {
            String text = Files.readString(file, StandardCharsets.UTF_8);
            if (text.isBlank()) {
                return new LinkedHashMap<>();
            }
            return yaml.load(text);
        } catch (IOException e) {
            throw new EnvyException(List.of("reading " + file.getFileName() + ": " + e.getMessage()));
        } catch (RuntimeException e) {
            throw new EnvyException(List.of("parsing " + file.getFileName() + ": " + e.getMessage()));
        }
    }

    private static Path findUpward(String name, Path start) {
        Path dir = start.normalize();
        for (;;) {
            Path candidate = dir.resolve(name);
            if (Files.isRegularFile(candidate)) {
                return candidate;
            }
            Path parent = dir.getParent();
            if (parent == null || parent.equals(dir)) {
                return null;
            }
            dir = parent;
        }
    }

    private static String currentBranch(Path repoDir) {
        try {
            Process process = new ProcessBuilder(
                    "git", "-C", repoDir.toString(), "rev-parse", "--abbrev-ref", "HEAD")
                    .redirectError(ProcessBuilder.Redirect.DISCARD)
                    .start();
            String out = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8).trim();
            int exit = process.waitFor();
            if (exit == 0 && !out.isEmpty() && !"HEAD".equals(out)) {
                return out;
            }
            return null;
        } catch (IOException | InterruptedException e) {
            return null;
        }
    }

    private static String sanitizeBranch(String branch) {
        StringBuilder sb = new StringBuilder(branch.length());
        for (char c : branch.toCharArray()) {
            sb.append(Character.isLetterOrDigit(c) || c == '.' || c == '_' || c == '-' ? c : '-');
        }
        return sb.toString();
    }

    private static String scalarToString(Object value, String key, List<String> problems) {
        if (value instanceof String s) {
            return s;
        }
        if (value instanceof Boolean || value instanceof Number) {
            return String.valueOf(value);
        }
        problems.add(key + ": expected a scalar value");
        return null;
    }

    private static Boolean asBool(Object value) {
        return value instanceof Boolean b ? b : Boolean.parseBoolean(String.valueOf(value));
    }

    private static void checkType(Map<String, Object> spec, String key, String raw, List<String> problems) {
        String type = String.valueOf(spec.getOrDefault("type", "string"));
        switch (type) {
            case "integer":
                if (!raw.trim().matches("-?\\d+")) {
                    problems.add(key + ": expected an integer, got \"" + raw + "\"");
                }
                break;
            case "number":
            case "float":
                try {
                    Double.parseDouble(raw.trim());
                } catch (NumberFormatException e) {
                    problems.add(key + ": expected a number, got \"" + raw + "\"");
                }
                break;
            case "boolean":
            case "bool":
                if (!List.of("true", "false", "1", "0", "yes", "no", "on", "off")
                        .contains(raw.trim().toLowerCase())) {
                    problems.add(key + ": expected a boolean (true/false), got \"" + raw + "\"");
                }
                break;
            default:
                break;
        }
    }

    private static void checkFormat(Map<String, Object> spec, String key, String raw, List<String> problems) {
        Object formatObj = spec.get("format");
        if (formatObj == null) {
            return;
        }
        String format = String.valueOf(formatObj);
        switch (format) {
            case "uri":
            case "url": {
                int idx = raw.indexOf("://");
                if (idx <= 0 || raw.contains(" ")) {
                    problems.add(key + ": does not satisfy format '" + format + "': \"" + raw + "\"");
                    return;
                }
                String scheme = raw.substring(0, idx);
                if (!KNOWN_SCHEMES.contains(scheme) && nearMissScheme(scheme)) {
                    String suggestion = bestScheme(scheme);
                    if (suggestion != null) {
                        problems.add(key + ": does not satisfy format '" + format
                                + "' — did you mean " + suggestion + "://" + raw.substring(idx + 3) + "?");
                    }
                }
                break;
            }
            case "email": {
                int at = raw.indexOf('@');
                boolean ok = at > 0
                        && raw.indexOf('.', at) > at + 1
                        && !raw.contains(" ");
                if (!ok) {
                    problems.add(key + ": does not satisfy format 'email': \"" + raw + "\"");
                }
                break;
            }
            case "uuid":
                if (!raw.matches("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}")) {
                    problems.add(key + ": does not satisfy format 'uuid': \"" + raw + "\"");
                }
                break;
            default:
                break;
        }
    }

    private static boolean nearMissScheme(String scheme) {
        return bestScheme(scheme) != null;
    }

    private static String bestKeyMatch(String key, Set<String> candidates) {
        String best = null;
        int bestDistance = Integer.MAX_VALUE;
        for (String candidate : candidates) {
            int distance = osaDistance(key, candidate);
            int limit = Math.max(1, Math.min(3, key.length() / 3));
            if (distance <= limit && distance < bestDistance) {
                best = candidate;
                bestDistance = distance;
            }
        }
        return best;
    }

    private static String bestScheme(String scheme) {
        String best = null;
        int bestDistance = Integer.MAX_VALUE;
        for (String candidate : KNOWN_SCHEMES) {
            int distance = osaDistance(scheme, candidate);
            int limit = Math.max(1, Math.min(3, scheme.length() / 3));
            if (distance <= limit && distance < bestDistance) {
                best = candidate;
                bestDistance = distance;
            }
        }
        return best;
    }

    private static String mockValue(String key) {
        StringBuilder hex = new StringBuilder();
        long hash = 0xcbf29ce484222325L;
        for (int i = 0; i < key.length(); i++) {
            hash ^= key.charAt(i);
            hash *= 0x100000001b3L;
        }
        hex.append(String.format("%016x", hash)).append(String.format("%016x", hash >>> 1));
        return "mock_" + hex;
    }

    private static int osaDistance(String a, String b) {
        int[][] d = new int[a.length() + 1][b.length() + 1];
        for (int i = 0; i <= a.length(); i++) {
            d[i][0] = i;
        }
        for (int j = 0; j <= b.length(); j++) {
            d[0][j] = j;
        }
        for (int i = 1; i <= a.length(); i++) {
            for (int j = 1; j <= b.length(); j++) {
                int cost = a.charAt(i - 1) == b.charAt(j - 1) ? 0 : 1;
                d[i][j] = Math.min(Math.min(d[i - 1][j] + 1, d[i][j - 1] + 1), d[i - 1][j - 1] + cost);
                if (i > 1 && j > 1
                        && a.charAt(i - 1) == b.charAt(j - 2)
                        && a.charAt(i - 2) == b.charAt(j - 1)) {
                    d[i][j] = Math.min(d[i][j], d[i - 2][j - 2] + 1);
                }
            }
        }
        return d[a.length()][b.length()];
    }
}

File: packages/php/src/Envy.php
<?php

declare(strict_types=1);

namespace Envy;

use Symfony\Component\Yaml\Yaml;

/**
 * Loads and validates {@code envy.yaml} natively in PHP — Laravel, Symfony,
 * plain scripts, anything. Mirrors the precedence of the envy core engine:
 *
 *   putenv/env → envy.local.<branch>.yaml → envy.local.yaml → schema default → generated mock
 *
 * Usage:
 *   $config = \Envy\Envy::load();
 *   $dbUrl  = $config['DATABASE_URL'];
 */
final class Envy
{
    private const SCHEMA_FILE = 'envy.yaml';

    private const KNOWN_SCHEMES = [
        'postgresql', 'postgres', 'mysql', 'mariadb', 'mssql', 'mongodb',
        'mongodb+srv', 'redis', 'rediss', 'amqp', 'rabbitmq', 'kafka',
        'http', 'https', 'ws', 'wss', 'grpc', 'ftp', 'sftp', 'ssh',
        'smtp', 's3', 'gs', 'azblob', 'sqlite',
    ];

    /** @var array<string, array<string, string>>|null */
    private static ?array $cache = null;

    private function __construct()
    {
    }

    /** Load from the current working directory (searching upward). */
    public static function load(): array
    {
        return self::loadFrom(getcwd() ?: '.');
    }

    /**
     * Load searching upward from $startDir. Cached per schema location.
     *
     * @return array<string, string> resolved variable => value
     */
    public static function loadFrom(string $startDir): array
    {
        $schemaPath = self::findUpward(self::SCHEMA_FILE, self::normalize($startDir));
        if ($schemaPath === null) {
            throw new EnvyException([
                sprintf('no %s found in %s or any parent directory — run `envy init` first', self::SCHEMA_FILE, $startDir),
            ]);
        }

        if (self::$cache !== null && isset(self::$cache[$schemaPath])) {
            return self::$cache[$schemaPath];
        }

        $problems = [];
        $schemaDoc = self::readYaml($schemaPath, $problems);
        if ($schemaDoc === null || !is_array($schemaDoc)) {
            throw new EnvyException([$schemaPath.': empty or invalid schema']);
        }

        /** @var array<string, array<string, mixed>> $config */
        $config = is_array($schemaDoc['config'] ?? null) ? $schemaDoc['config'] : [];

        $baseDir = dirname($schemaPath);
        $branch = self::currentBranch($baseDir);
        $overlayFile = $branch !== null
            ? $baseDir.DIRECTORY_SEPARATOR.'envy.local.'.self::sanitizeBranch($branch).'.yaml'
            : null;
        $localFile = $baseDir.DIRECTORY_SEPARATOR.'envy.local.yaml';

        $overlay = ($overlayFile !== null && is_file($overlayFile))
            ? self::valuesOf(self::readYaml($overlayFile, $problems), basename($overlayFile), $problems)
            : [];
        $local = is_file($localFile)
            ? self::valuesOf(self::readYaml($localFile, $problems), 'envy.local.yaml', $problems)
            : [];

        $resolved = [];

        foreach ($config as $key => $spec) {
            $spec = is_array($spec) ? $spec : [];
            $placed = null;
            $value = null;

            $envValue = getenv($key);
            if ($envValue !== false && $envValue !== '') {
                $value = $envValue;
                $placed = true;
            } elseif (array_key_exists($key, $overlay)) {
                $value = self::scalarToString($overlay[$key], $key, $problems);
                $placed = $value !== null;
            } elseif (array_key_exists($key, $local)) {
                $value = self::scalarToString($local[$key], $key, $problems);
                $placed = $value !== null;
            } elseif (array_key_exists('default', $spec) && $spec['default'] !== null) {
                $value = self::scalarToString($spec['default'], $key, $problems);
                $placed = $value !== null;
            } elseif (!empty($spec['mock'])) {
                $value = self::mockValue($key);
                $placed = true;
            }

            if (!$placed || $value === null) {
                if (!empty($spec['required'])) {
                    $problems[] = 'missing required variable '.$key;
                }
                continue;
            }

            self::checkType($spec, $key, $value, $problems);
            self::checkFormat($spec, $key, $value, $problems);
            $resolved[$key] = $value;
        }

        foreach (['branch overlay' => $overlay, 'envy.local.yaml' => $local] as $label => $layer) {
            foreach (array_keys($layer) as $key) {
                if (!array_key_exists($key, $config)) {
                    $suggestion = self::bestKeyMatch((string) $key, array_keys($config));
                    $problems[] = $key.' is set in '.$label.' but not declared in '
                        .self::SCHEMA_FILE.' (typo?)'
                        .($suggestion !== null ? ' — did you mean '.$suggestion.'?' : '');
                }
            }
        }

        if ($problems !== []) {
            throw new EnvyException($problems);
        }

        self::$cache[$schemaPath] = $resolved;

        return $resolved;
    }

    /** Convenience accessor over load(). */
    public static function get(string $key): ?string
    {
        return self::load()[$key] ?? null;
    }

    // ---------------------------------------------------------------- internals

    private static function normalize(string $dir): string
    {
        $real = realpath($dir);

        return $real === false ? rtrim($dir, '/\\') : $real;
    }

    private static function findUpward(string $name, string $start): ?string
    {
        $dir = $start;
        for (;;) {
            $candidate = $dir.DIRECTORY_SEPARATOR.$name;
            if (is_file($candidate)) {
                return $candidate;
            }
            $parent = dirname($dir);
            if ($parent === $dir) {
                return null;
            }
            $dir = $parent;
        }
    }

    private static function readYaml(string $file, array &$problems): mixed
    {
        try {
            $text = file_get_contents($file);
            if ($text === false || trim((string) $text) === '') {
                return null;
            }

            return Yaml::parse($text);
        } catch (\Throwable $e) {
            $problems[] = basename($file).': '.$e->getMessage();

            return null;
        }
    }

    /**
     * Accepts the nested `values:` form or a flat SCREAMING_SNAKE mapping.
     *
     * @return array<string, mixed>
     */
    private static function valuesOf(mixed $doc, string $label, array &$problems): array
    {
        if ($doc === null) {
            return [];
        }
        if (!is_array($doc)) {
            $problems[] = $label.': expected a mapping';

            return [];
        }
        if (isset($doc['values']) && is_array($doc['values'])) {
            /** @var array<string, mixed> $values */
            $values = $doc['values'];

            return $values;
        }
        unset($doc['service'], $doc['version']);
        foreach (array_keys($doc) as $key) {
            if (!is_string($key) || $key !== strtoupper($key) || str_contains($key, ':')) {
                $problems[] = $label.': expected a `values:` mapping or flat KEY: value pairs';

                return [];
            }
        }

        /** @var array<string, mixed> $doc */
        return $doc;
    }

    private static function scalarToString(mixed $value, string $key, array &$problems): ?string
    {
        if (is_string($value)) {
            return $value;
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_int($value) || is_float($value)) {
            return (string) $value;
        }
        $problems[] = $key.': expected a scalar value';

        return null;
    }

    private static function currentBranch(string $repoDir): ?string
    {
        $cmd = 'git -C '.escapeshellarg($repoDir).' rev-parse --abbrev-ref HEAD 2>&1';
        $out = @shell_exec($cmd);
        $branch = is_string($out) ? trim($out) : '';
        if ($branch === '' || $branch === 'HEAD' || str_contains($branch, 'fatal')) {
            return null;
        }

        return $branch;
    }

    private static function sanitizeBranch(string $branch): string
    {
        return preg_replace('/[^A-Za-z0-9._-]/', '-', $branch) ?? 'unknown';
    }

    private static function checkType(array $spec, string $key, string $raw, array &$problems): void
    {
        $type = is_string($spec['type'] ?? null) ? $spec['type'] : 'string';
        $trimmed = trim($raw);
        switch ($type) {
            case 'integer':
                if (preg_match('/^-?\d+$/', $trimmed) !== 1) {
                    $problems[] = sprintf('%s: expected an integer, got "%s"', $key, $raw);
                }
                break;
            case 'number':
            case 'float':
                if (is_numeric($trimmed) === false) {
                    $problems[] = sprintf('%s: expected a number, got "%s"', $key, $raw);
                }
                break;
            case 'boolean':
            case 'bool':
                if (!in_array(strtolower($trimmed), ['true', 'false', '1', '0', 'yes', 'no', 'on', 'off'], true)) {
                    $problems[] = sprintf('%s: expected a boolean (true/false), got "%s"', $key, $raw);
                }
                break;
            default:
                break;
        }
    }

    private static function checkFormat(array $spec, string $key, string $raw, array &$problems): void
    {
        $format = $spec['format'] ?? null;
        if (!is_string($format)) {
            return;
        }
        switch ($format) {
            case 'uri':
            case 'url':
                $idx = strpos($raw, '://');
                if ($idx === false || $idx === 0 || preg_match('/\s/', $raw) === 1) {
                    $problems[] = sprintf("%s: does not satisfy format '%s': \"%s\"", $key, $format, $raw);
                    break;
                }
                $scheme = substr($raw, 0, $idx);
                if (!in_array($scheme, self::KNOWN_SCHEMES, true)) {
                    $best = self::bestScheme($scheme);
                    if ($best !== null) {
                        $problems[] = sprintf(
                            "%s: does not satisfy format '%s' — did you mean %s://%s?",
                            $key,
                            $format,
                            $best,
                            substr($raw, $idx + 3)
                        );
                    }
                }
                break;
            case 'email':
                $at = strpos($raw, '@');
                $ok = $at !== false && $at > 0 && strpos($raw, '.', $at + 1) > $at + 1
                    && preg_match('/\s/', $raw) !== 1;
                if (!$ok) {
                    $problems[] = sprintf("%s: does not satisfy format 'email': \"%s\"", $key, $raw);
                }
                break;
            case 'uuid':
                if (preg_match('/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/', $raw) !== 1) {
                    $problems[] = sprintf("%s: does not satisfy format 'uuid': \"%s\"", $key, $raw);
                }
                break;
            default:
                break;
        }
    }

    private static function bestKeyMatch(string $key, array $candidates): ?string
    {
        $limit = max(1, min(3, intdiv(strlen($key), 3)));
        $best = null;
        $bestDistance = PHP_INT_MAX;
        foreach ($candidates as $candidate) {
            $distance = self::osaDistance($key, (string) $candidate);
            if ($distance <= $limit && $distance < $bestDistance) {
                $best = (string) $candidate;
                $bestDistance = $distance;
            }
        }

        return $best;
    }

    private static function bestScheme(string $scheme): ?string
    {
        $limit = max(1, min(3, intdiv(strlen($scheme), 3)));
        $best = null;
        $bestDistance = PHP_INT_MAX;
        foreach (self::KNOWN_SCHEMES as $candidate) {
            $distance = self::osaDistance($scheme, $candidate);
            if ($distance <= $limit && $distance < $bestDistance) {
                $best = $candidate;
                $bestDistance = $distance;
            }
        }

        return $best;
    }

    /** Optimal string alignment distance (catches ture→true style typos). */
    private static function osaDistance(string $a, string $b): int
    {
        $m = strlen($a);
        $n = strlen($b);
        $d = [];
        for ($i = 0; $i <= $m; $i++) {
            $d[$i][0] = $i;
        }
        for ($j = 0; $j <= $n; $j++) {
            $d[0][$j] = $j;
        }
        for ($i = 1; $i <= $m; $i++) {
            for ($j = 1; $j <= $n; $j++) {
                $cost = $a[$i - 1] === $b[$j - 1] ? 0 : 1;
                $d[$i][$j] = min($d[$i - 1][$j] + 1, $d[$i][$j - 1] + 1, $d[$i - 1][$j - 1] + $cost);
                if ($i > 1 && $j > 1 && $a[$i - 1] === $b[$j - 2] && $a[$i - 2] === $b[$j - 1]) {
                    $d[$i][$j] = min($d[$i][$j], $d[$i - 2][$j - 2] + 1);
                }
            }
        }

        return $d[$m][$n];
    }

    private static function mockValue(string $key): string
    {
        $hash = 0xcbf29ce484222325;
        for ($i = 0, $len = strlen($key); $i < $len; $i++) {
            $hash ^= ord($key[$i]);
            $hash = ($hash * 0x100000001b3) & 0xFFFFFFFFFFFFFFFF;
        }

        return 'mock_'.sprintf('%016x%016x', $hash, $hash >> 1);
    }
}
